File: leetcode/js/merge-similar-items.js
/**
 * @param {number[][]} items1
 * @param {number[][]} items2
 * @return {number[][]}
 */
var mergeSimilarItems = function (items1, items2) {
  const itemsMap = new Map(items1);

  for (const [v, w] of items2) {
    itemsMap.set(v, itemsMap.has(v) ? itemsMap.get(v) + w : w);
  }

  const result = [...itemsMap];

  result.sort((a, b) => a[0] - b[0]);

  return result;
};

File: leetcode/rust/merge-similar-items.rs
impl Solution {
    pub fn merge_similar_items(items1: Vec<Vec<i32>>, items2: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        use std::collections::HashMap;

        let mut items_map = HashMap::new();

        for item in items1.iter().chain(items2.iter()) {
            let (v, w) = (item[0], item[1]);

            items_map
                .entry(v)
                .and_modify(|weight| *weight += w)
                .or_insert(w);
        }

        let mut result: Vec<Vec<i32>> = items_map.into_iter().map(|(v, w)| vec![v, w]).collect();

        result.sort_by_key(|item| item[0]);

        result
    }
}
